common/cache/metricsScopeCache.go

```diff
@@ -26,47 +26,96 @@ import (
 	"bytes"
 	"strconv"
 	"sync"
+	"sync/atomic"
+	"time"
 
 	"github.com/uber/cadence/common/metrics"
 )
 
+const flushBufferedMetricsScopeDuration = 10 * time.Second
+
 type domainMetricsScopeCache struct {
-	sync.RWMutex
 	scopeMap map[string]metrics.Scope
+	buffer   *buffer
+}
+
+type buffer struct {
+	sync.RWMutex
+	bufferMap map[string]metrics.Scope
 }
 
+type metricsScopeMap map[string]metrics.Scope
+
+var cache atomic.Value
+
 // NewDomainMetricsScopeCache constructs a new domainMetricsScopeCache
 func NewDomainMetricsScopeCache() DomainMetricsScopeCache {
-	return &domainMetricsScopeCache{
+	mc := &domainMetricsScopeCache{
 		scopeMap: make(map[string]metrics.Scope),
+		buffer: &buffer{
+			bufferMap: make(map[string]metrics.Scope),
+		},
+	}
+	cache.Store(make(metricsScopeMap))
+	go mc.flushBufferedMetricsScope()
+	return mc
+}
+
+func (c *domainMetricsScopeCache) flushBufferedMetricsScope() {
+	for {
+		select {
+		case <-time.After(flushBufferedMetricsScopeDuration):
```

Comment:
May worth adding a unit test for the case where scopes are added to the cache in different iterations of this flush loop (to make sure previously flushed scope are still there)